### PR TITLE
Sygus Lambda Grammars

### PR DESCRIPTION
**Pull request by ajreynol**
This pull request changes all non-builtin sygus operators so they are represented as lambda terms, fixing #1344.  

This includes a few different cases for user-provided grammars:

- Lambdas are used to avoid duplicates of builtin operators, for example:
A -> A + A | B + B 
B -> ...
We use the sygus constructors c1( A, A ) with sygus operator + and c2( B, B ) with sygus operator(lambda (xy) (+ x y)) for A.

- Defined functions are replaced by lambdas, for example: 
(define-fun inc ((x Int)) Int (+ x 1))
A -> inc( B )
B -> ...
We use the sygus constructor c( B ) with sygus operator (lambda (x) (inc x)) for A.  Notice that the body of the lambda of the sygus constructor is of the form (APPLY inc x).  It remains to expand definitions for symbols occurring in lambdas (see #1304). 
Also notice that the case of 0-argument defined functions:
(define-fun val ((x Int)) Int 5)
A -> v
We use the sygus constructor c() with sygus operator (v), internally represented as (APPLY v).

- Let expressions (see printer/sygus_print_callback.h for details).

It also fixes a minor issue discovered in ce_guided_single_inv.cpp regarding variable ordering.

### CODE REVIEW

File: src/parser/smt2/smt2.cpp [hunk 1]
```diff
@@ -926,23 +927,15 @@ void Smt2::mkSygusDatatype( CVC4::Datatype& dt, std::vector<CVC4::Expr>& ops,
...
-  std::vector< std::vector<Expr> > df_let_args;
-  std::vector< Expr > df_let_body;
-  //dt.mkSygusConstructors( ops, cnames, cargs, sygus_to_builtin, 
-  //                        d_sygus_let_func_to_vars, d_sygus_let_func_to_body, d_sygus_let_func_to_num_input_vars, 
-  //                        df_name, df_op, df_let_args, df_let_body );
   
   Debug("parser-sygus") << "SMT2 sygus parser : Making constructors for sygus datatype " << dt.getName() << std::endl;
   Debug("parser-sygus") << "  add constructors..." << std::endl;
-  for( int i=0; i<(int)cnames.size(); i++ ){
+  for (unsigned i = 0; i < cnames.size(); i++)
```
Comment (HanielB) [inline src/parser/smt2/smt2.cpp]: no ``size()`` in condition

File: src/parser/smt2/smt2.cpp [hunk 2]
```diff
@@ -959,60 +952,122 @@ void Smt2::mkSygusDatatype( CVC4::Datatype& dt, std::vector<CVC4::Expr>& ops,
...
+        // callback prints as the expression
+        spc = std::make_shared<printer::SygusExprPrintCallback>(body, largs);
       }
-      //make the let_body
-      std::vector< Expr > children;
-      if( ops[i].getKind() != kind::BUILTIN ){
-        children.push_back( ops[i] );
+      else
+      {
+        std::map<CVC4::Expr, CVC4::Expr>::iterator it =
```
Comment (HanielB) [inline src/parser/smt2/smt2.cpp]: You are in CVC4 namespace, remove the prefix.

File: src/parser/smt2/smt2.h [hunk 3]
```diff
@@ -319,6 +319,19 @@ class Smt2 : public Parser {
...
 
+  /** make sygus bound var list
+   *
+   * This is used for converting non-builtin sygus operators to lambda
+   * expressions. It takes as input a datatype and constructor index (for
+   * naming) and a vector of type ltypes.
+   * It appends a bound variable to lvars for each type in ltypes, and returns
+   * a bound variable list whose children are lvars.
+   */
+  CVC4::Expr makeSygusBoundVarList(CVC4::Datatype& dt,
```
Comment (HanielB) [inline src/parser/smt2/smt2.h]: ditto

File: src/parser/smt2/smt2.h [hunk 4]
```diff
@@ -319,6 +319,19 @@ class Smt2 : public Parser {
...
+                                   unsigned i,
+                                   const std::vector<CVC4::Type>& ltypes,
+                                   std::vector<CVC4::Expr>& lvars);
```
Comment (HanielB) [inline src/parser/smt2/smt2.h]: ditto

File: src/theory/quantifiers/ce_guided_conjecture.cpp [hunk 5]
```diff
@@ -428,8 +430,9 @@ void CegConjecture::getModelValues( std::vector< Node >& n, std::vector< Node >&
       TypeNode tn = nv.getType();
       Trace("cegqi-engine") << n[i] << " -> ";
       std::stringstream ss;
-      std::vector< Node > lvs;
-      TermDbSygus::printSygusTerm( ss, nv, lvs );
+      // std::vector< Node > lvs;
+      // TermDbSygus::printSygusTerm( ss, nv, lvs );
```
Comment (HanielB) [inline src/theory/quantifiers/ce_guided_conjecture.cpp]: remove commented code

File: src/theory/quantifiers/ce_guided_conjecture.cpp [hunk 6]
```diff
@@ -621,8 +624,9 @@ void CegConjecture::printSynthSolution( std::ostream& out, bool singleInvocation
       if( status==0 ){
         out << sol;
       }else{
-        std::vector< Node > lvs;
-        TermDbSygus::printSygusTerm( out, sol, lvs );
+        // std::vector< Node > lvs;
+        // TermDbSygus::printSygusTerm( out, sol, lvs );
```
Comment (HanielB) [inline src/theory/quantifiers/ce_guided_conjecture.cpp]: ditto

File: src/theory/quantifiers/quantifiers_attributes.h [hunk 7]
```diff
@@ -55,6 +55,17 @@ typedef expr::Attribute< SygusAttributeId, bool > SygusAttribute;
 struct SynthesisAttributeId {};
 typedef expr::Attribute< SynthesisAttributeId, bool > SynthesisAttribute;
 
+/** Attribute for setting printing information for sygus variables
+ * For variable d of sygus datatype type, if
```
Comment (HanielB) [inline src/theory/quantifiers/quantifiers_attributes.h]: newline

File: src/printer/sygus_print_callback.h [hunk 8]
```diff
@@ -12,18 +12,61 @@
...
+
+ protected:
+  /** body of the sygus term */
+  Expr d_body;
+  /** arguments */
+  std::vector<Expr> d_args;
+  /** body argument */
+  int d_body_argument;
+  /** do string replace
+   * Replaces all occurrences of oldStr with newStr in str.
```
Comment (HanielB) [inline src/printer/sygus_print_callback.h]: newline

**Comment by ajreynol** [comment]
I've addressed Haniel's review in my latest commit.

**Pull request merged**